File: Cpp/1414_Find_the_Minimum_Number_of_Fibonacci_Numbers_Whose_Sum_Is_K.cpp
class Solution {
public:
    int findMinFibonacciNumbers(int k) {
        int result = 0;
        vector<int> fi(2, 1);
        
        for(int i=2; i<=k && fi[i-1]<k; ++i){
            fi.push_back(fi[i-1] + fi[i-2]);
        }
        
        int n = k;
        for(int i=fi.size()-1; i>=0; --i){
            if(n-fi[i]>=0){
                n-=fi[i];
                ++result;
            }
        }
        
        return result;
    }
};

File: Kotlin/1414_Find_the_Minimum_Number_of_Fibonacci_Numbers_Whose_Sum_Is_K.kt
class Solution {
    fun findMinFibonacciNumbers(k: Int): Int {
        var result = 0
        val fi = mutableListOf<Int>()
        
        fi.add(1)
        fi.add(1)
        
        for(i in 2 .. k){
            if(fi[i - 1] >= k) break
            fi.add(fi[i - 1] + fi[i - 2])
        }
        
        var n = k
        for(i in fi.size - 1 downTo 1){
            if(n - fi[i] < 0) continue
            
            n -= fi[i]
            ++result
        }
        
        return result
    }
}